cli/client.go

```diff
@@ -883,6 +888,111 @@ func (c *AppClient) UpdateModule(manifest ModuleManifest, organizationID *string
 	return c.client.UpdateModule(c.c.Context, &req)
 }
 
+// UploadModuleFile wraps the grpc UploadModuleFile request.
+func (c *AppClient) UploadModuleFile(
+	moduleID,
+	version,
+	platform string,
+	organizationID *string,
+	file *os.File,
+) (*apppb.UploadModuleFileResponse, error) {
+	if err := c.ensureLoggedIn(); err != nil {
+		return nil, err
+	}
+	ctx := c.c.Context
+
+	stream, err := c.client.UploadModuleFile(ctx)
+	if err != nil {
+		return nil, err
+	}
+	moduleFileInfo := apppb.ModuleFileInfo{
+		ModuleId:       moduleID,
+		OrganizationId: organizationID,
+		Version:        version,
+		Platform:       platform,
+	}
+	req := &apppb.UploadModuleFileRequest{
+		ModuleFile: &apppb.UploadModuleFileRequest_ModuleFileInfo{ModuleFileInfo: &moduleFileInfo},
+	}
+	if err := stream.Send(req); err != nil {
+		return nil, err
+	}
+
+	var errs error
+	// We do not add the EOF as an error because all server-side errors trigger and EOF on the stream
+	// This results in extra clutter to the error msg
+	if err := sendModuleUploadRequests(ctx, stream, file, c.c.App.Writer); err != nil && !errors.Is(err, io.EOF) {
+		errs = multierr.Combine(errs, errors.Wrapf(err, "error uploading %s", file.Name()))
+	}
+
+	resp, closeErr := stream.CloseAndRecv()
+	errs = multierr.Combine(errs, closeErr)
+	return resp, errs
+}
+
+func sendModuleUploadRequests(ctx context.Context, stream apppb.AppService_UploadModuleFileClient, f *os.File, stdout io.Writer) error {
+	stat, err := f.Stat()
+	if err != nil {
+		return err
+	}
+	fileSize := stat.Size()
+	uploadedBytes := 0
+	// Close the line with the progress reading
+	defer fmt.Fprint(stdout, "\n")
+
+	//nolint:errcheck
+	defer stream.CloseSend()
+	// Loop until there is no more content to be read from file.
+	for {
+		select {
+		case <-ctx.Done():
+			return context.Canceled
+		default:
+			// Get the next UploadRequest from the file.
+			uploadReq, err := getNextModuleUploadRequest(ctx, f)
+
+			// EOF means we've completed successfully.
+			if errors.Is(err, io.EOF) {
+				return nil
+			}
+
+			if err != nil {
+				return errors.Wrap(err, "file reading error")
+			}
+
+			if err = stream.Send(uploadReq); err != nil {
+				return err
+			}
+			uploadedBytes += len(uploadReq.GetFile())
+			// Simple progress reading until we have a proper tui library
+			uploadPercent := int(math.Ceil(100 * float64(uploadedBytes) / float64(fileSize)))
+			fmt.Fprintf(stdout, "\r\aUploading... %d%% (%d/%d bytes)", uploadPercent, uploadedBytes, fileSize)
```

Comment:
nit: i wonder if this should be a better generic function to print to stdout. Something like PrintToStout(string x) so we don't have random fmt.printf statements everywhere.

Reply:
Agreed -- it is terrible. However, that is the pattern set by the rest of the `cli` and I don't feel like deviating or creating a new print function just because it is a bit verbose.

All of the print functions will have to be replaced in the future during the productize the cli phase when we get a `tui` library that has niceties like debug levels and colored text.

I am in favor of waiting until then for people in the future to make an informed decision about how to print to stdout

Reply:
can you add a ticket for this? Yeah, I get that its probs generic to how we do stuff for the CLI now but yeah doesn't feel right.

Reply:
https://viam.atlassian.net/browse/RSDK-4076